File: frontend/src/api/servers.ts
import { api } from "./client";

export type ServerHealthStatus = "unknown" | "online" | "offline" | "degraded";

export type ServerRead = {
  id: number;
  name: string;
  provider_type: "xui";
  country: string;
  location: string | null;
  panel_url: string;
  enabled: boolean;
  max_users: number | null;
  current_users: number;
  priority: number;
  subscription_base_url: string | null;
  last_health_status: ServerHealthStatus;
  last_health_checked_at: string | null;
  created_at: string;
  updated_at: string;
};

export type ServerCreate = {
  name: string;
  country: string;
  location?: string;
  panel_url: string;
  username?: string;
  password?: string;
  api_token?: string;
  enabled: boolean;
  max_users?: number;
  priority: number;
  subscription_base_url?: string;
};

export type ServerUpdate = Partial<ServerCreate>;

export type ServerHealthRead = {
  ok: boolean;
  status: ServerHealthStatus;
  message: string;
};

export type ServerInboundRead = {
  id: number;
  remark: string | null;
  protocol: string | null;
  enable: boolean | null;
  port: number | null;
};

export type XuiClientRead = {
  inbound_id: number;
  inbound_remark: string | null;
  protocol: string | null;
  email: string;
  client_uuid: string | null;
  sub_id: string | null;
  enable: boolean | null;
  expiry_time: number | null;
  traffic_limit: number;
  up: number;
  down: number;
  total: number;
};

export async function listServers(): Promise<ServerRead[]> {
  const response = await api.get<ServerRead[]>("/servers");
  return response.data;
}

export async function createServer(payload: ServerCreate): Promise<ServerRead> {
  const response = await api.post<ServerRead>("/servers", payload);
  return response.data;
}

export async function updateServer(
  serverId: number,
  payload: ServerUpdate,
): Promise<ServerRead> {
  const response = await api.patch<ServerRead>(`/servers/${serverId}`, payload);
  return response.data;
}

export async function checkServer(serverId: number): Promise<ServerHealthRead> {
  const response = await api.post<ServerHealthRead>(`/servers/${serverId}/check`);
  return response.data;
}

export async function listServerInbounds(serverId: number): Promise<ServerInboundRead[]> {
  const response = await api.get<ServerInboundRead[]>(`/servers/${serverId}/inbounds`);
  return response.data;
}

export async function listXuiClients(serverId: number): Promise<XuiClientRead[]> {
  const response = await api.get<XuiClientRead[]>(`/servers/${serverId}/xui-clients`);
  return response.data;
}

File: app/schemas/servers.py
from datetime import datetime

from pydantic import BaseModel, Field, HttpUrl

from app.core.enums import PanelProviderType, ServerHealthStatus


class ServerCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    country: str = Field(min_length=1, max_length=128)
    location: str | None = None
    panel_url: HttpUrl
    username: str | None = None
    password: str | None = None
    api_token: str | None = None
    enabled: bool = True
    max_users: int | None = Field(default=None, ge=1)
    priority: int = Field(default=100, ge=0)
    subscription_base_url: HttpUrl | None = None


class ServerUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    country: str | None = Field(default=None, min_length=1, max_length=128)
    location: str | None = None
    panel_url: HttpUrl | None = None
    username: str | None = None
    password: str | None = None
    api_token: str | None = None
    enabled: bool | None = None
    max_users: int | None = Field(default=None, ge=1)
    priority: int | None = Field(default=None, ge=0)
    subscription_base_url: HttpUrl | None = None


class ServerRead(BaseModel):
    id: int
    name: str
    provider_type: PanelProviderType
    country: str
    location: str | None
    panel_url: str
    enabled: bool
    max_users: int | None
    current_users: int
    priority: int
    subscription_base_url: str | None
    last_health_status: ServerHealthStatus
    last_health_checked_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ServerHealthRead(BaseModel):
    ok: bool
    status: ServerHealthStatus
    message: str


class ServerInboundRead(BaseModel):
    id: int
    remark: str | None = None
    protocol: str | None = None
    enable: bool | None = None
    port: int | None = None


class XuiClientRead(BaseModel):
    inbound_id: int
    inbound_remark: str | None = None
    protocol: str | None = None
    email: str
    client_uuid: str | None = None
    sub_id: str | None = None
    enable: bool | None = None
    expiry_time: int | None = None
    traffic_limit: int = 0
    up: int = 0
    down: int = 0
    total: int = 0
